api-admin/src/main/java/bio/terra/pearl/api/admin/service/auth/EnforcePortalEnrolleePermissionAspect.java

```diff
@@ -0,0 +1,88 @@
+package bio.terra.pearl.api.admin.service.auth;
+
+import bio.terra.pearl.api.admin.service.auth.context.PortalEnrolleeAuthContext;
+import bio.terra.pearl.core.model.participant.Enrollee;
+import bio.terra.pearl.core.model.portal.Portal;
+import bio.terra.pearl.core.model.study.PortalStudy;
+import bio.terra.pearl.core.model.study.StudyEnvironment;
+import bio.terra.pearl.core.service.exception.NotFoundException;
+import bio.terra.pearl.core.service.participant.EnrolleeService;
+import bio.terra.pearl.core.service.study.StudyEnvironmentService;
+import lombok.extern.slf4j.Slf4j;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.aspectj.lang.reflect.MethodSignature;
+import org.springframework.stereotype.Component;
+
+/**
+ * Aspect to enforce portal permissions on methods annotated with EnforcePortalStudyEnvPermission.
+ */
+@Component
+@Aspect
+@Slf4j
+public class EnforcePortalEnrolleePermissionAspect
+    extends BaseEnforcePermissionAspect<
+        PortalEnrolleeAuthContext, EnforcePortalEnrolleePermission> {
+  private final AuthUtilService authUtilService;
+  private final StudyEnvironmentService studyEnvironmentService;
+  private final EnrolleeService enrolleeService;
+
+  public EnforcePortalEnrolleePermissionAspect(
+      AuthUtilService authUtilService,
+      StudyEnvironmentService studyEnvironmentService,
+      EnrolleeService enrolleeService) {
+    this.authUtilService = authUtilService;
+    this.studyEnvironmentService = studyEnvironmentService;
+    this.enrolleeService = enrolleeService;
+  }
+
+  @Around(value = "@annotation(EnforcePortalEnrolleePermission)")
+  public Object enforcePermission(ProceedingJoinPoint joinPoint) throws Throwable {
+    PortalEnrolleeAuthContext authContext = extractAuthContext(joinPoint);
+    String permission = getPermissionName(joinPoint);
+    Portal portal =
+        authUtilService.authUserToPortalWithPermission(
+            authContext.getOperator(), authContext.getPortalShortcode(), permission);
+    authContext.setPortal(portal);
+
+    PortalStudy portalStudy =
+        authUtilService.authUserToStudy(
+            authContext.getOperator(),
+            authContext.getPortalShortcode(),
+            authContext.getStudyShortcode());
+    authContext.setPortalStudy(portalStudy);
+
+    StudyEnvironment studyEnv =
+        studyEnvironmentService.verifyStudy(
+            authContext.getStudyShortcode(), authContext.getEnvironmentName());
+    authContext.setStudyEnvironment(studyEnv);
+
+    Enrollee enrollee =
+        enrolleeService
+            .findByShortcodeAndStudyEnvId(authContext.getEnrolleeShortcode(), studyEnv.getId())
+            .orElseThrow(() -> new NotFoundException("Enrollee not found"));
+
+    authContext.setEnrollee(enrollee);
```

Comment:
lots of duplicated code with EnforcePortalStudyEnvPermissionAspect, but I'm not sure the best way to make it reusable

Reply:
agreed -- I spent a while trying to dedupe stuff across the existing aspects, and couldn't figure it out because of the way these are based on annotations, which are interfaces, and therefore can't inherit from each other.